game-app/game-core/src/main/java/games/strategy/triplea/odds/calculator/BattleCalculatorPanel.java

```diff
@@ -1111,6 +1095,16 @@
           }
         }
 
+        // Mimic the legacy behavior and force the parties to be non-allied modifying either
+        // attacker or defender like the old code did). This legacy behavior was requested in #9532.
+        if (data.getRelationshipTracker().isAllied(getDefender(), getAttacker())) {
+          if (location.isWater() && players.isEmpty()) {
+            getEnemy(getAttacker()).ifPresent(this::setDefender);
+          } else {
+            getEnemy(getDefender()).ifPresent(this::setAttacker);
```

Comment:
Can you clarify why the battle location and players being empty or not would change whether we need to set the 'defender' or 'attacker'?

Reply:
Sure, that's what the old code did.  If you asking why the old code chose those conditions and why they chose to alter attacker over defender or vice versa, then you have to ask the person who initially wrote that.

Reply:
Walking through this in a debugger, I think the comment no longer agrees with the if/else statement. Essentially the logic is says:
- if we have not found two opposing factions
  - if we are water, then the attacker is correct and find an enemy faction
  - else, defender is correct and find an enemy faction
Basically this seems to be an extension of the previous logic that is assigning an attacker and a defending player. Hence, it's odd to have a comment stating that we are going to choose two opposing sides here to keep with a users expectations when the previous 30 lines of code are also doing exactly that but for different cases. There is an implication that this could never have been removed and/or the preceding logic is dead code.

Does it make sense why I say the comment and this if statement do not jive together?

I think we are also violating the DRY principle where we do the same thing in multiple places. Notably if we assign an attacker then the same assignment earlier would be a dead store. Seemingly this if block needs to be merged with preceding code for it to make more sense.

Reply:
> Basically this seems to be an extension of the previous logic that is assigning an attacker and a defending player.

It's not an extension, the same thing was also done by the old code but was well hidden behind a chain of EventListeners and was removed in this PR:

https://github.com/triplea-game/triplea/blob/6fd988dfa42fd7e46e0660276599aa6fa496d81b/game-app/game-core/src/main/java/games/strategy/triplea/odds/calculator/BattleCalculatorPanel.java#L959-L961


> Basically this seems to be an extension of the previous logic that is assigning an attacker and a defending player. Hence, it's odd to have a comment stating that we are going to choose two opposing sides here to keep with a users expectations when the previous 30 lines of code are also doing exactly that but for different cases. There is an implication that this could never have been removed and/or the preceding logic is dead code.
>
> Does it make sense why I say the comment and this if statement do not jive together?

Sorry, can't follow you there.  I think maybe the confusion arises from the fact that the code change in the event listener affected this logic?  Since this code is now gone, we have to move parts of the event listener logic here in the setup code.

> I think we are also violating the DRY principle where we do the same thing in multiple places.

Yes, it does but moving this code into the above stacked if-cases would duplicate this if three times (we cannot join the isWater() and isEmpty() case).  I contemplated about it but decided for 2 as it is smaller than 3.  Essentially we have a 3x2 grid guarded by ifs and I don't think we can avoid that.

> Notably if we assign an attacker then the same assignment earlier would be a dead store.

I can understand your hesitation about dead store but

a) it is already dead store, e.g. attacker get assigned (potentially) multiple times:
   - https://github.com/triplea-game/triplea/blob/6fd988dfa42fd7e46e0660276599aa6fa496d81b/game-app/game-core/src/main/java/games/strategy/triplea/odds/calculator/BattleCalculatorPanel.java#L1075
   - https://github.com/triplea-game/triplea/blob/6fd988dfa42fd7e46e0660276599aa6fa496d81b/game-app/game-core/src/main/java/games/strategy/triplea/odds/calculator/BattleCalculatorPanel.java#L1086  (jepp, that one did in the old code reset the attacker via the event listener)
   - https://github.com/triplea-game/triplea/blob/6fd988dfa42fd7e46e0660276599aa6fa496d81b/game-app/game-core/src/main/java/games/strategy/triplea/odds/calculator/BattleCalculatorPanel.java#L1089
   - https://github.com/triplea-game/triplea/blob/6fd988dfa42fd7e46e0660276599aa6fa496d81b/game-app/game-core/src/main/java/games/strategy/triplea/odds/calculator/BattleCalculatorPanel.java#L1095  (dito)
   - https://github.com/triplea-game/triplea/blob/6fd988dfa42fd7e46e0660276599aa6fa496d81b/game-app/game-core/src/main/java/games/strategy/triplea/odds/calculator/BattleCalculatorPanel.java#L1099  (dito)
   - https://github.com/triplea-game/triplea/blob/6fd988dfa42fd7e46e0660276599aa6fa496d81b/game-app/game-core/src/main/java/games/strategy/triplea/odds/calculator/BattleCalculatorPanel.java#L1101

b) I cannot see how to make a single assignment for attacker/defender in all cases, e.g. the land case would become
```java
if (!location.isWater()) {
        defenderCombo.setSelectedItem(location.getOwner());
        boolean setAttacker = false;
        for (final GamePlayer player : players) {
                if (Matches.isAtWar(getDefender(), data.getRelationshipTracker()).test(player)) {
                        attackerCombo.setSelectedItem(player);
                        setAttacker = true;
                        break;
                }
        }
        if (!setAttacker) {
                // Currently player has priority if at war.
                if (currentPlayer.isPresent() && Matches.isAtWar(getDefender(), data.getRelationshipTracker()).test(currentPlayer.get())) {
                                attackerCombo.setSelectedItem(currentPlayer.get());
                } else {
                        // find an enemy if possible
                        Optional<Player> enemy = getEnemy(getDefender());
                        if (enemy.isPresent()) {
                                attackerCombo.setSelectedItem(enemy.get());
                        } else {
                                // Fall back to current player (they are allied but there are no enemies anyway)
                                if (currentPlayer.isPresent()) {
                                        attackerCombo.setSelectedItem(currentPlayer.get());
                                } else {
                                        attackerCombo.setSelectedItem(getFirstPlayer();
                                }
                        }
                }
        }
} else {...}
```

(pseudo code might not compile)

So what should we do?

Reply:
Thank you for the response. To answer your question, we either spend quite a bit more time refactoring to get maybe cleaner code, or we live with where we are. I am concerned by both scope creep and ROI if we spend a lot more time now, on this. So, having been given more context, I think we can move on and call things good enough. I re-read the diff and it's quite reasonable.